Feat/rpc methods WIP (#71)
kadena-io/KIPs: kip-0039.md

Category: Interface
Created: 2025-02-17
---

# kadena_getNetwork_v1 API

## Abstract

This KIP standardizes the `kadena_getNetwork_v1` method for retrieving information about the currently active network from a Kadena wallet. The method provides consistent network details to dApps, ensuring they can correctly adjust to the appropriate blockchain environment.

## Motivation

A uniform approach to obtaining network information simplifies dApp development by ensuring that network details are provided in a consistent format across different wallet implementations. This standardization will:

- Provide reliable network details (such as network name and networkId) to dApps.
- Simplify integration and reduce potential mismatches between wallet and dApp network settings.
- Enhance the overall user experience by ensuring clear network context.

## Specification

### Method Definition

**Method Name:** `kadena_getNetwork_v1`

### Request Parameters

```typescript
interface GetNetworkRequest {
  // No parameters are required for this method.
}
```

### Response

The method returns a NetworkInfo object containing details about the active network:

```typescript
interface NetworkInfo {

>   name: string // The human-readable name of the network.

Reviewer: ```suggestion
  networkName: string // The human-readable name of the network.
```
Use `networkName` as key for this?

  networkId: string // The unique identifier for the network (e.g., "mainnet01").

>   url?: string // Optional: The network endpoint URL.

Reviewer: Specify what this endpoint represents. Whether it should be just `api.chainweb.com` or the full path to where the commands can be sent eg.: `https://api.chainweb.com/chainweb/0.0/mainnet01/chain/1/pact/api/v1`.
Should this contain placeholders for chainId?

Reviewer 2: I would say consistency is key so I agree.

>   url?: string // Optional: The network endpoint URL.

Reviewer 3: url can be a list of nodes as well.

Author: @alber70g javad brings up the same thing we thought about, but we decided to have a single url, want to re-discuss?

Author: @barthuijgen

}
```

### Error Responses

In case of failure, the wallet MUST return an error according to the [jsonrpc spec](https://www.jsonrpc.org/specification#error_object) which represent the following type:

```typescript
interface ErrorResponse {
  id: number
  jsonrpc: '2.0'
  error: {
    code: number
    message: string
    data?: object
  }
}
```

### Error types

| Code   | Message               |
| ------ | --------------------- |
| -32603 | Internal server error |

### Behavior Requirements

1. Network Information Retrieval:

   - MUST return accurate and up-to-date information about the currently active network.
   - MUST support all networks that the wallet is configured to operate with.

2. Error Handling:
   - MUST return a clear error if network information cannot be retrieved due to configuration issues or unsupported networks.

# Examples

## Basic Get Network Request

### Request

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "kadena_getNetwork_v1",
  "params": {}
}
```

### Success Response

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {

>     "name": "kadena:mainnet01",

Reviewer: I wouldn't use `kadena:` in this. The RPC will never return `ethereum:xyz` so it's kinda redundant.
I would consider `networkName` as the display name to show the user in the UI

```suggestion
    "name": "Mainnet",
```

---
KIP: "0039"
Title: kadena_getNetwork_v1 API
Author: Danillo Felixdaal, Bart Huijgen and Albert Groothedde
Status: Draft
Type: Standard
Category: Interface
Created: 2025-02-17
---

# kadena_getNetwork_v1 API

## Abstract

This KIP standardizes the `kadena_getNetwork_v1` method for retrieving
information about the **currently active network** from a Kadena wallet. The
method provides consistent network details to dApps, ensuring they can correctly
adjust to the appropriate blockchain environment.

## Motivation

A uniform approach to obtaining network information simplifies dApp development
by ensuring that network details are provided in a consistent format across
different wallet implementations. This standardization will:

- Provide reliable network details (such as network name and networkId) to
  dApps.
- Simplify integration and reduce potential mismatches between wallet and dApp
  network settings.
- Enhance the overall user experience by ensuring clear network context.

> The dapp can use the URL to make /local calls from the network.

Reviewer 4: This is the first mention of `URL` in the document. This needs an transition sentence between the previous points (benefits of this method) to the current point (dApps can use the URL).

Maybe something like:
```suggestion
The wallet will also return a list of endpoint URLs for the active network. The dApp can use these URLs to make `/local` calls to the network.
```

and/or move this entire URL section after the Response type definition in the `Specification` section.

(similar comment as in KIP 40)

> The wallet returns only the host. The dapp should append networkId and chainId
> to the URL.

Reviewer 4: ```suggestion
The wallet returns only the host. The dApp needs to append the `networkId` and `chainId` to the URL to be able to communicate with the network.
```

Example:

- Wallet returns: https://api.testnet.chainweb.com
- Dapp can send requests to:
  https://api.testnet.chainweb.com/chainweb/0.0/testnet04/chain/19/pact/api/v1/local

> [!NOTE]  
> For more information about Chainweb API see

> > [Chainweb API Documentation](api.chainweb.com)

Reviewer 4: Not redirecting to actual website
```suggestion
> [Chainweb API Documentation](https://api.chainweb.com/openapi/)
```

## Specification

### Method Definition

**Method Name:** `kadena_getNetwork_v1`

### Response

> The method returns a NetworkInfo object containing details about the active

Reviewer 4: ```suggestion
The method returns a `NetworkInfo` object containing details about the active
```
